diag_manager/fms_diag_object_container.F90

```diff
@@ -0,0 +1,201 @@
+         riter = this%iterator() !!An iterator over the entire container
+      endif
+      do while( riter%has_data() .eqv. .true.)
+         tdo = riter%get()
+         if(id == tdo%get_id() ) then
+            EXIT
+         end if
+         status = riter%next()
+      end do
+   end function find_diag_object
+
+   !> \brief insert diagnostic object obj with given id.
+   ! Objects are inserted at the back / end of the list
+   ! This version of the container also enforces that the
+   ! objects ID is equal the input key. status of -1
+   ! is returned fro errors
+   !! TODO: what should be the reutrn type
+   !!TODO: WE ASSUME id == obj%diagID
+   function insert_diag_object (this, id, obj) result (status)
+      class (fms_diag_object_container), intent (in out) :: this
+      integer,  intent (in) :: id
+      class(fms_diag_object) , intent (in out) :: obj
+      integer :: status
+      !!
+      class(literator), allocatable ::  tliter
+
+      status = -1
+      if ( id .ne. obj%get_id() ) then
+         !!TODO: log error
+      endif
+      tliter =  this%the_ll%push_back( obj )
+      if(tliter%has_data() .eqv. .true. ) then
+         status = 0
+      endif
+   end function
+
+   !> \brief Remove and return the first object in the container with the corresponding key .
+   ! Return an iterator that starts with the next object on the container. (Note, if the client
+   ! code does not already have a referenc to the object being removed, then  the client may
+   ! want to to use subroutine find before using subroutine remove. If subroutine find is used,
+   ! consider then calling remove with the iterator returned from find.)
+   function remove_diag_object (this, key, iiter ) result (riter)
+      class (fms_diag_object_container), intent (in out) :: this
+      integer , intent (in) :: key
+      class(diag_obj_iterator), intent (in), optional :: iiter
+      class(diag_obj_iterator), allocatable:: riter
+      class(literator), allocatable :: temp_liter
+      integer :: status
+      !!
+      !Find the object in the container.
+      riter = this%find ( key , iiter)
+      temp_liter = this%the_ll%remove( riter%liter%current )
+      riter = diag_obj_iterator(temp_liter)
+   end function
+
+   !> return the number of objects held in the container
+   function get_num_objects (this )result (sz)
+      class (fms_diag_object_container), intent (in out) :: this
+      integer :: sz
+      sz = this%the_ll%size()
+   end function
+
+
+   !> return a pointer to an iterator for the objects in the continer.
+   function get_iterator (this) result (oliter)
+      class (fms_diag_object_container), intent (in) :: this
+      class(diag_obj_iterator) , allocatable :: oliter
+      oliter = diag_obj_iterator( this%the_ll%get_literator() )
+   end function
+
+
+   function diag_obj_iterator_constructor( iliter ) result (diag_itr)
+    class (literator), allocatable :: iliter
+    class (diag_obj_iterator), allocatable :: diag_itr
```

Comment:
Variables need to be documented in doxygen style

Reply:
done